common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/ClientCertAuthChallengeHandler.java

```diff
@@ -496,43 +605,91 @@ public void run() {
                 //Update Dialog to indicate that an incorrect attempt was made.
-                synchronized (sSmartcardDialogLock) {
-                    ((SmartcardPinDialog) mCurrentDialog).setErrorMode();
-                }
+                dialogHolder.setPinDialogErrorMode();
             }
         }
     }
 
     /**
-     * Sets up and shows dialog when user has exceeded the max attempts allowed to enter their YubiKey PIN.
+     * Attempts to authenticate using smartcard certificate.
+     * @param cert Chosen certificate for authentication.
+     * @param pin char array containing PIN.
+     * @param slotAlias Name of slot where certificate is located.
+     * @param piv A PivSession created from a SmartCardConnection that can interact with certificates located in the PIV slots on the YubiKey.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
      */
-    private void showMaxAttemptsDialog(){
-        final SmartcardMaxFailedAttemptsDialog maxFailedAttemptsDialog = new SmartcardMaxFailedAttemptsDialog(
-                new SmartcardMaxFailedAttemptsDialog.PositiveButtonListener() {
-                    @Override
-                    public void onClick() {
-                        //Reset currentDialog to null
-                        synchronized (sSmartcardDialogLock) {
-                            mCurrentDialog = null;
-                        }
-                    }
-                },
-                mActivity);
-        //set current dialog to maxFailedAttemptsDialog and show.
-        synchronized (sSmartcardDialogLock) {
-            mCurrentDialog = maxFailedAttemptsDialog;
-            maxFailedAttemptsDialog.show();
+    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+    private void useSmartcardCertForAuth(@NonNull final X509Certificate cert,
+                                         @NonNull final char[] pin,
+                                         @NonNull final String slotAlias,
+                                         @NonNull final PivSession piv,
+                                         @NonNull final ClientCertRequest request) {
+        final String methodTag = TAG + "useSmartcardCertForAuth:";
+        //Need to add a PivProvider instance to the beginning of the array of Security providers in order for signature logic to occur.
+        //Note that this provider is removed when the UsbYubiKeyDevice connection is closed.
+        Security.insertProviderAt(new PivProvider(getPivProviderCallback(request)), 1);
+        try {
+            //Using KeyStore methods in order to generate PivPrivateKey.
+            //Loading null is needed for initialization.
+            final KeyStore keyStore = KeyStore.getInstance("YKPiv", new PivProvider(piv));
+            keyStore.load(null);
+            //PivPrivateKey implements PrivateKey
+            final PivPrivateKey pivPrivateKey = (PivPrivateKey) keyStore.getKey(slotAlias, pin);
```

Comment:
Can we run into ClassCastException here? Should we be checking type first?

Reply:
I'll add a type check, just in case. If the key retrieved is not of type PivPrivateKey, an error will be logged, a generic error dialog will be displayed, and the request will be cancelled.